ics_space_amplification_goal_test.py

```diff
@@ -0,0 +1,232 @@
+        InfoEvent(message=f"Altered table by: {full_alter_query}")
+
+    def test_ics_space_amplification_goal(self):  # pylint: disable=too-many-locals
+        """
+        (1) writing new data. wait for compactions to finish.
+        (2) over-writing existing data.
+        (3) measure space amplification after over-writing with SAG=None,1.5,1.2,None
+        """
+
+        # (1) writing new data.
+        prepare_write_cmd = self.params.get('prepare_write_cmd')
+        InfoEvent(message=f"Starting C-S prepare load: {prepare_write_cmd}")
+        self.run_prepare_write_cmd()
+        InfoEvent(message="Wait for compactions to finish after write is done.")
+        self.wait_no_compactions_running()
+
+        stress_cmd = self.params.get('stress_cmd', default=None)
+        sag_testing_values = [None, '1.5', '1.2', '1.5', None]
+        column_size = 200
+        num_of_columns = 5
+        # TODO: find the right number, corresponding to yaml stress_cmd
+        ops_num = overwrite_ops_num = 1000200300  # TODO: for debug only: should be x5 bigger: 200200300 = 200GB *5 = 1TB
+        total_data_to_overwrite_gb = round(overwrite_ops_num * column_size * num_of_columns / (1024 ** 3), 2)
+
+        # (2) over-writing existing data.
+        for sag in sag_testing_values:
+            dict_nodes_capacity_before_overwrite_data = self._get_nodes_used_capacity()
+            InfoEvent(
+                message=f"Nodes used capacity before start overwriting data: {dict_nodes_capacity_before_overwrite_data}")
+            additional_compaction_params = None if not sag else {'space_amplification_goal': sag}
+            # (3) Altering compaction with SAG=None,1.5,1.2,None
+            # TODO: verify how to remove previous SAG value
+            self._alter_table_compaction(additional_compaction_params=additional_compaction_params)
+            stress_queue = list()
+            InfoEvent(message=f"Starting C-S over-write load: {stress_cmd}")
+
+            start_time = time.time()
+            params = {'keyspace_num': 1, 'stress_cmd': stress_cmd,
+                      'round_robin': self.params.get('round_robin')}
+            self._run_all_stress_cmds(stress_queue, params)
+
+            for stress in stress_queue:
+                self.verify_stress_thread(cs_thread_pool=stress)
+
+            InfoEvent(message="Wait for compactions to finish after over-write is done.")
+            self.wait_no_compactions_running()
+            # (3) measure space amplification for the re-written data
+            self.measure_nodes_space_amplification_after_write(
+                dict_nodes_initial_capacity=dict_nodes_capacity_before_overwrite_data,
+                written_data_size_gb=total_data_to_overwrite_gb, start_time=start_time)
+
+        InfoEvent(message=f"Space-amplification-goal testing cycles are done.")
+
+
+def generate_node_capacity_query_postfix(node):
+    node_ip = node.private_ip_address  # due to pylint bug
```

Comment:
which bug is that ?

Reply:
perhaps a legacy issue. in anyway it's an existing master code..